src/libexpr/search-path.hh

```diff
@@ -0,0 +1,93 @@
+#pragma once
+///@file
+
+#include <optional>
+
+#include "types.hh"
+#include "comparator.hh"
+
+namespace nix {
+
+/**
+ * A single element of a `SearchPath`.
+ */
+struct SearchPathElem
+{
+    /**
+     * What we are looking up, first element of a path
+     */
+    typedef std::string Key;
+
+    /**
+     * Where we look, a path or a URL (with certain restrictions / extra
+     * structure).
+     *
+     * @todo Maybe change this to `std::variant<SourcePath, URL>`.
+     */
+    typedef std::string Value;
+
+    /**
+     * `key=value` entries.
+     *
+     * The key must match exactly
+     */
+    struct KeyValue {
+        Key key;
+        Value value;
+
+        GENERATE_CMP(KeyValue, me->key, me->value);
+    };
+
+    /**
+     * `value` entries.
+     *
+     * `value` must resolve to a directory and we search inside for the
+     * key.
+     */
+    struct JustValue {
+        Value value;
+
+        GENERATE_CMP(JustValue, me->value);
+    };
```

Comment:
I learned that `foo/bar=baz` actually, works, so arguably

- the key is a last of `/` separated compoents
- `foo` is sugar for `=foo`
- `JustValue` is a special case of `KeyValue`

And we don't need the `std::variant` I introduced.

On the flip side, I am not sure whether what I learned is an intentional feature :sweat_smile:. If it is not, we could give it a deprecation warning at some point and what I wrote is probably better after all because it better reflects the intent.